contrib/slurm/slurmexecutor.py

```diff
@@ -0,0 +1,324 @@
+        self.run_finished_callback = run_finished_callback
+        self.benchmark = benchmark
+        self.output_handler = output_handler
+        self.setDaemon(True)
+
+        self.start()
+
+    def run(self):
+        while not STOPPED_BY_INTERRUPT:
+            try:
+                currentRun = _Worker.working_queue.get_nowait()
+            except queue.Empty:
+                return
+
+            try:
+                logging.debug('Executing run "%s"', currentRun.identifier)
+                self.execute(currentRun)
+                logging.debug('Finished run "%s"', currentRun.identifier)
+            except SystemExit as e:
+                logging.critical(e)
+            except BenchExecException as e:
+                logging.critical(e)
+            except BaseException:
+                logging.exception("Exception during run execution")
+            self.run_finished_callback()
+            _Worker.working_queue.task_done()
+
+    def execute(self, run):
+        """
+        This function executes the tool with a sourcefile with options.
+        It also calls functions for output before and after the run.
+        """
+        self.output_handler.output_before_run(run)
+
+        args = run.cmdline()
+        logging.debug("Command line of run is %s", args)
+
+        try:
+            with open(run.log_file, "w") as f:
+                for i in range(6):
+                    f.write(os.linesep)
+
+            run_result = run_slurm(
+                self.benchmark,
+                args,
+                run.log_file,
+            )
+
+        except KeyboardInterrupt:
+            # If the run was interrupted, we ignore the result and cleanup.
+            stop()
+
+        if STOPPED_BY_INTERRUPT:
+            try:
+                if self.benchmark.config.debug:
+                    os.rename(run.log_file, run.log_file + ".killed")
+                else:
+                    os.remove(run.log_file)
+            except OSError:
+                pass
+            return 1
+
+        run.set_result(run_result)
+        self.output_handler.output_after_run(run)
+        return None
+
+
+jobid_pattern = re.compile(r"job (\d*) queued")
+
+
+def run_slurm(benchmark, args, log_file):
+    timelimit = benchmark.rlimits.cputime
+    cpus = benchmark.rlimits.cpu_cores
+    memory = benchmark.rlimits.memory
+
+    srun_timelimit_h = int(timelimit / 3600)
+    srun_timelimit_m = int((timelimit % 3600) / 60)
+    srun_timelimit_s = int(timelimit % 60)
+    srun_timelimit = f"{srun_timelimit_h}:{srun_timelimit_m}:{srun_timelimit_s}"
+
+    mem_per_cpu = int(memory / cpus / 1000000)
+
+    assert benchmark.config.scratchdir and os.path.exists(benchmark.config.scratchdir)
+
+    with tempfile.TemporaryDirectory(dir=benchmark.config.scratchdir) as tempdir:
+
+        os.makedirs(os.path.join(tempdir, "upper"))
+        os.makedirs(os.path.join(tempdir, "work"))
+
+        singularity_command = (
+            (
+                [
```

Comment:
I would suggest to get rid of this deep indentation and at the same time construct the command in the intuitive order (left to right), by something like the following:
```python
command = ['srun', ...]
if benchmark.config.singularity:
    singularity_command = [`singularity`, ...]
    command.extend(singularity_command)
command.extend(args)
```

Reply:
sure, that's a good idea.